File: x86_trace_generator/pin/source/tools/SimpleExamples/malloctrace.cpp
/*
 * Copyright (C) 2004-2023 Intel Corporation.
 * SPDX-License-Identifier: MIT
 */

/*! @file
 *  This file contains an ISA-portable PIN tool for tracing calls to malloc
 */

#include "pin.H"
#include <iostream>
#include <fstream>







/* ===================================================================== */
/* Names of malloc and free */
/* ===================================================================== */

#define MALLOC "malloc"
#define FREE "free"

/* ===================================================================== */
/* Global Variables */
/* ===================================================================== */

std::ofstream TraceFile;

/* ===================================================================== */
/* Commandline Switches */
/* ===================================================================== */

KNOB< std::string > KnobOutputFile(KNOB_MODE_WRITEONCE, "pintool", "o", "malloctrace.out", "specify trace file name");

/* ===================================================================== */
/* Print Help Message                                                    */
/* ===================================================================== */

INT32 Usage()
{
    std::cerr << "This tool produces a trace of calls to malloc.\n"
            "\n";

    std::cerr << KNOB_BASE::StringKnobSummary();

    std::cerr << std::endl;

    return -1;
}

/* ===================================================================== */

VOID Arg1Before(CHAR* name, ADDRINT size) { TraceFile << name << "(" << size << ")" << std::endl; }

/* ===================================================================== */

VOID MallocAfter(ADDRINT ret) { TraceFile << "  returns " << ret << std::endl; }

/* ===================================================================== */

VOID Image(IMG img, VOID* v)
{
    RTN mallocRtn = RTN_FindByName(img, MALLOC);
    if (RTN_Valid(mallocRtn))
    {
        RTN_Open(mallocRtn);
        RTN_InsertCall(mallocRtn, IPOINT_BEFORE, (AFUNPTR)Arg1Before, IARG_ADDRINT, MALLOC, IARG_FUNCARG_ENTRYPOINT_VALUE, 0,
                       IARG_END);
        RTN_InsertCall(mallocRtn, IPOINT_AFTER, (AFUNPTR)MallocAfter, IARG_FUNCRET_EXITPOINT_VALUE, IARG_END);
        RTN_Close(mallocRtn);
    }

    RTN freeRtn = RTN_FindByName(img, FREE);
    if (RTN_Valid(freeRtn))
    {
        RTN_Open(freeRtn);
        RTN_InsertCall(freeRtn, IPOINT_BEFORE, (AFUNPTR)Arg1Before, IARG_ADDRINT, FREE, IARG_FUNCARG_ENTRYPOINT_VALUE, 0,
                       IARG_END);
        RTN_Close(freeRtn);
    }
}

/* ===================================================================== */

VOID Fini(INT32 code, VOID* v) { TraceFile.close(); }

/* ===================================================================== */
/* Main                                                                  */
/* ===================================================================== */

int main(int argc, char* argv[])
{
    PIN_InitSymbols();

    if (PIN_Init(argc, argv))
    {
        return Usage();
    }

    TraceFile.open(KnobOutputFile.Value().c_str());

    TraceFile << std::hex;
    TraceFile.setf(std::ios::showbase);

    std::cout << std::hex;
    std::cout.setf(std::ios::showbase);

    IMG_AddInstrumentFunction(Image, 0);
    PIN_AddFiniFunction(Fini, 0);

    // Never returns
    PIN_StartProgram();

    return 0;
}

/* ===================================================================== */
/* eof */
/* ===================================================================== */

File: x86_trace_generator/pin/bindings/glue/pinglueserver/tests/Scripts/malloctrace.py
#
# Copyright (C) 2024-2024 Intel Corporation.
# SPDX-License-Identifier: MIT
#

import pin

output_file = open("script_malloctrace.out", "w")
	
def Arg1Before(name, size):
    size = hex(size) if size else 0
    output_file.write(f"{name}({size})\n")

def MallocAfter(ret):
	output_file.write(f"  returns {hex(ret)}\n")

def image_instrumentation_cb(img):
    # Instrument the malloc() and free() functions.  Print the input argument
    # of each malloc() or free(), and the return value of malloc().
    # Find the malloc() function.
    mallocRtn = pin.RTN_FindByName(img, "malloc")
    if(pin.RTN_Valid(mallocRtn)):
        pin.RTN_Open(mallocRtn)
        # Instrument malloc() to print the input argument value and the return value.
        pin.RTN_InsertCall(mallocRtn, pin.IPOINT_BEFORE, Arg1Before, pin.IARG_PYOBJ, "malloc", pin.IARG_FUNCARG_ENTRYPOINT_VALUE, 0)
        pin.RTN_InsertCall(mallocRtn, pin.IPOINT_AFTER, MallocAfter, pin.IARG_FUNCRET_EXITPOINT_VALUE)

        pin.RTN_Close(mallocRtn)
	
    # Find the free() function.
    freeRtn = pin.RTN_FindByName(img, "free")
    if (pin.RTN_Valid(freeRtn)):
        pin.RTN_Open(freeRtn)
        # Instrument free() to print the input argument value.
        pin.RTN_InsertCall(freeRtn, pin.IPOINT_BEFORE, Arg1Before, pin.IARG_PYOBJ, "free", pin.IARG_FUNCARG_ENTRYPOINT_VALUE, 0)
        pin.RTN_Close(freeRtn)


pin.IMG_AddInstrumentFunction(image_instrumentation_cb)
